File: ios/App/History/HistorySessionsError.swift
// ---------------------------------------------------------------------------
// HistorySessionsError — the two sessions-load feedback affordances for the
// History panel, mirroring the webui sessions drawer (drawer.tsx):
//
//   - Empty-list failure  → SessionsErrorEmpty: a centered "Couldn't load —
//     Retry" message + Retry button, shown WHERE the list would be when the
//     fetch failed and no rows are loaded.
//   - Stale rows           → SessionsStaleBanner: "Showing saved results" and
//     refresh detail + Retry/checking, shown ABOVE the still-rendered rows.
//
// Both are stateless leaves: the host (HistorySidePanel) decides which to show
// from HistoryViewModel.error/loading + whether rows exist, and passes `onRetry`.
// State hoisting per the swiftui rule; no ViewModel reference.
//
// accessibilityIdentifiers: sessions-error-retry (empty), sessions-stale-banner
// (stale banner), sessions-stale-retry (stale action).
// ---------------------------------------------------------------------------
import SwiftUI

/// Centered empty-state shown in place of the list when the load failed and no
/// rows are present. Mirrors drawer.tsx's `EMPTY_LOAD_FAIL` + inline Retry.
struct SessionsErrorEmpty: View {
    let onRetry: () -> Void

    private static let message = "Couldn't load past chats."
    private static let retryCta = "Retry"

    var body: some View {
        VStack(spacing: Space.md) {
            Text(Self.message)
                .font(Typo.ui(TypeScale.sm))
                .foregroundStyle(DuskColors.ink3)
                .multilineTextAlignment(.center)
            DesignActionButton(
                title: Self.retryCta,
                role: .secondary,
                accessibilityId: "sessions-error-retry",
                fillsWidth: false,
                action: onRetry
            )
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, Space.xxl)
        .padding(.horizontal, Space.lg)
        .accessibilityElement(children: .contain)
    }
}

/// Banner shown above saved rows while a refresh is in flight or after that
/// refresh fails. The host derives `checking` from HistoryViewModel.loading;
/// this leaf owns no retry or cached-data state.
struct SessionsStaleBanner: View {
    let onRetry: () -> Void
    var checking = false

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    private static let title = "Showing saved results"
    private static let detail = "Couldn’t refresh just now."
    private static let retryCta = "Retry"
    private static let checkingCta = "Checking…"

    var body: some View {
        HStack(alignment: .center, spacing: Space.sm) {
            VStack(alignment: .leading, spacing: 0) {
                Text(Self.title)
                    .font(Typo.ui(DesignMetrics.controlLabelSize, .semibold))
                    .foregroundStyle(DuskColors.ink)
                Text(Self.detail)
                    .font(Typo.ui(TypeScale.sm))
                    .foregroundStyle(DuskColors.ink2)
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            DesignActionButton(
                title: checking ? Self.checkingCta : Self.retryCta,
                role: .quiet,
                state: checking ? .disabled : .normal,
                accessibilityId: "sessions-stale-retry",
                fillsWidth: false,
                action: onRetry
            )
        }
        .padding(.horizontal, Space.md)
        .padding(.vertical, Space.sm)
        .frame(minHeight: StaleStyle.minimumHeight)
        .background {
            SessionsStaleMaterialCanvas(
                face: staleBackground,
                topBorder: staleTopBorder
            )
        }
        .designPlate()
        .animation(
            DesignV2.Motion.animation(duration: DesignV2.Motion.state, reduceMotion: reduceMotion),
            value: checking
        )
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("sessions-stale-banner")
    }

    private var staleBackground: Color {
        checking
            ? DuskColors.bgElev.overlaying(DuskColors.accent50, opacity: StaleStyle.checkingBackgroundMix)
            : DuskColors.bgElev.overlaying(DuskColors.amber, opacity: StaleStyle.restBackgroundMix)
    }

    private var staleTopBorder: Color {
        checking
            ? DuskColors.lineSoft.overlaying(DuskColors.accent, opacity: StaleStyle.checkingBorderMix)
            : DuskColors.lineSoft.overlaying(DuskColors.amber, opacity: StaleStyle.restBorderMix)
    }
}

/// The stale state has a reviewed, History-specific tint layered inside the
/// shared plate. One Canvas owns that face and directional top rule so the
/// product state does not rebuild the plate from stacked SwiftUI decorations.
private struct SessionsStaleMaterialCanvas: View {
    let face: Color
    let topBorder: Color

    var body: some View {
        Canvas { context, size in
            let rect = CGRect(origin: .zero, size: size)
            let shape = RoundedRectangle(cornerRadius: Radii.md, style: .continuous)
            context.fill(shape.path(in: rect), with: .color(face))
            context.fill(
                Path(CGRect(
                    x: 0,
                    y: DesignMetrics.hairline,
                    width: size.width,
                    height: DesignMetrics.hairline
                )),
                with: .color(topBorder)
            )
        }
        .allowsHitTesting(false)
        .accessibilityHidden(true)
    }
}

private enum StaleStyle {
    // The source's 58pt banner grows to the action + vertical padding and
    // includes its top border plus the enclosing plate border.
    static let minimumHeight: CGFloat = 63
    static let restBackgroundMix = 0.08
    static let restBorderMix = 0.40
    static let checkingBackgroundMix = 0.24
    static let checkingBorderMix = 0.45
}

#Preview("Sessions error — empty") {
    ZStack {
        DuskColors.bg.ignoresSafeArea()
        SessionsErrorEmpty(onRetry: {})
    }
    .preferredColorScheme(.dark)
}

#Preview("Sessions error — stale banner") {
    ZStack(alignment: .top) {
        DuskColors.bg.ignoresSafeArea()
        SessionsStaleBanner(onRetry: {})
            .padding(.horizontal, Space.lg)
            .padding(.top, Space.lg)
    }
    .preferredColorScheme(.dark)
}

#Preview("Sessions error — stale banner checking") {
    ZStack(alignment: .top) {
        DuskColors.bg.ignoresSafeArea()
        SessionsStaleBanner(onRetry: {}, checking: true)
            .padding(.horizontal, Space.lg)
            .padding(.top, Space.lg)
    }
    .preferredColorScheme(.dark)
}

File: android/src/main/kotlin/io/sentient/android/history/HistorySessionsError.kt
// ---------------------------------------------------------------------------
// HistorySessionsError — the two sessions-load-failure affordances for the
// History drawer, mirroring the iOS HistorySessionsError (parity:
// ios/App/History/HistorySessionsError.swift) + the webui sessions drawer:
//
//   - Empty-list failure → SessionsErrorEmpty: a centered "Couldn't load — Retry"
//     message + Retry button, shown WHERE the list would be when the fetch failed
//     and no rows are loaded. testTag: sessions-error-retry.
//   - Stale failure      → SessionsStaleBanner: a thin "Sync failed — list may be
//     stale." banner + Retry, shown ABOVE the still-rendered (stale) rows.
//     testTag: sessions-stale-retry.
//
// Both are stateless leaves: the host (HistoryContent) decides which to show from
// the pure [HistoryUiState.showsErrorEmpty] / [HistoryUiState.showsStaleBanner]
// derivations and passes `onRetry`. State hoisting per the android-compose rule;
// no ViewModel reference.
// ---------------------------------------------------------------------------
package io.sentient.android.history

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import io.sentient.android.theme.LocalTokens
import io.sentient.android.theme.SentientTheme
import io.sentient.mobilesdk.design.Colors

private const val EMPTY_ERROR_MSG = "Couldn't load past chats."
private const val STALE_MSG = "Sync failed — list may be stale."
private const val RETRY_CTA = "Retry"
private val BANNER_BORDER = 1.dp

/**
 * Centered empty-state shown in place of the list when the load failed and no
 * rows are present. Mirrors the iOS SessionsErrorEmpty.
 */
@Composable
fun SessionsErrorEmpty(onRetry: () -> Unit, modifier: Modifier = Modifier) {
    val tokens = LocalTokens.current
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(vertical = tokens.space.xxl, horizontal = tokens.space.lg),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(tokens.space.md),
    ) {
        Text(
            text = EMPTY_ERROR_MSG,
            color = Color(Colors.ink3),
            fontSize = tokens.type.sm,
            textAlign = TextAlign.Center,
        )
        TextButton(onClick = onRetry, modifier = Modifier.testTag("sessions-error-retry")) {
            Text(
                text = RETRY_CTA,
                color = Color(Colors.accent),
                fontSize = tokens.type.sm,
                fontWeight = FontWeight.SemiBold,
            )
        }
    }
}

/**
 * Thin banner shown above the (stale) rows when a re-fetch failed but rows are
 * already loaded. Mirrors the iOS SessionsStaleBanner.
 */
@Composable
fun SessionsStaleBanner(onRetry: () -> Unit, modifier: Modifier = Modifier) {
    val tokens = LocalTokens.current
    Row(
        modifier = modifier
            .fillMaxWidth()
            .background(Color(Colors.bgElev), RoundedCornerShape(tokens.radii.md))
            .border(BANNER_BORDER, Color(Colors.lineSoft), RoundedCornerShape(tokens.radii.md))
            .padding(horizontal = tokens.space.md, vertical = tokens.space.sm),
        horizontalArrangement = Arrangement.spacedBy(tokens.space.sm),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = STALE_MSG,
            modifier = Modifier.weight(1f),
            color = Color(Colors.ink2),
            fontSize = tokens.type.xs,
            fontWeight = FontWeight.Medium,
        )
        TextButton(onClick = onRetry, modifier = Modifier.testTag("sessions-stale-retry")) {
            Text(
                text = RETRY_CTA,
                color = Color(Colors.ink),
                fontSize = tokens.type.xs,
                fontWeight = FontWeight.SemiBold,
            )
        }
    }
}

@Preview(name = "Sessions error — empty", showBackground = true, backgroundColor = 0xFF2B2621)
@Composable
private fun SessionsErrorEmptyPreview() {
    SentientTheme { SessionsErrorEmpty(onRetry = {}) }
}

@Preview(name = "Sessions error — stale", showBackground = true, backgroundColor = 0xFF2B2621)
@Composable
private fun SessionsStaleBannerPreview() {
    SentientTheme { SessionsStaleBanner(onRetry = {}) }
}
